simulators/main/event.py

```diff
@@ -0,0 +1,304 @@
+import json, logging
+from datetime import datetime
+from random import randint, uniform, choices
+from cumulocityAPI import CumulocityAPI
+from task import Task
+import interface
+
+cumulocityAPI = CumulocityAPI()
+log = logging.getLogger("events generation")
+
+
+class Event(interface.MachineType):
+    def __init__(self, model, shiftplans) -> None:
+        self.model = model
+        self.definitions = self.model.get('events', [])
+        self.shiftplans = shiftplans
+        self.machine_up = False
+        self.shutdown = False
+        self.production_time_s = 0.0
+        self.last_production_time_update = datetime.timestamp(datetime.utcnow())
+        self.out_of_production_time_logged = False
+        if self.model.get('enabled'):
+            self.production_speed_s = self.get_production_speed_s()
+
+    def callback(self, definition, min_interval_in_seconds, max_interval_in_seconds):
+        event_callback = lambda task: {Event.event_mapping[definition["type"]](self, definition, task)}
+        if definition:
+            log.debug(f'Machine {self.model.get("label")}, id {self.model.get("id")}: create periodic task for {definition["type"]}, interval in range ({min_interval_in_seconds}, {max_interval_in_seconds}) seconds')
+        else:
+            log.debug(f'No definition of event in machine {self.model.get("label")}, id {self.model.get("id")}')
+        return event_callback
+
+    def get_production_speed_s(self) -> float:
+        """Returns pieces/s"""
+        for event_definition in self.definitions:
+            event_type = event_definition.get("type") or ""
+            if event_type == "Piece_Produced":
+                frequency = event_definition.get("frequency")
+                return frequency / 3600.0
+            if event_type == "Pieces_Produced":
+                frequency = event_definition.get("frequency")
+                countMaximumPerHour = event_definition.get("countMaximumPerHour")
+                return frequency * countMaximumPerHour / 3600.0
+
+        return 0.0
+
+    def produce_pieces(self):
+        production_time = datetime.timestamp(datetime.utcnow()) - self.last_production_time_update
+        self.last_production_time_update = datetime.timestamp(datetime.utcnow())
+        if self.machine_up and not self.shutdown:
+            self.production_time_s = self.production_time_s + production_time
+            log.debug(f'{self.device_id} production time: {self.production_time_s}s')
```

Comment:
Is device_id present here?

Reply:
The device_ids are implemented in the model at the beginning so I can also say Yes. However, the IDE cannot recognize that because it is not implemented directly in the EVENT class but in the MachineSimulator class